ath79: TP-Link CPE210 v3

bluelineXY opened the pull request:
Adding Support for TP-Link CPE210 v3
Specifications:

    * SoC: Qualcomm Atheros QCA9533 (650MHz)
    * RAM: 64MB
    * Storage: 8 MB SPI NOR
    * Wireless: 2.4GHz N based built into SoC 2x2
    * Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

Installation:
    Flash factory image through stock firmware WEB UI or through TFTP
    To get to TFTP recovery just hold reset button while powering on for around 4-5 seconds and release.
    Rename factory image to recovery.bin
    Stock TFTP server IP:192.168.0.100
    Stock device TFTP adress:192.168.0.254

Thanks to robimarko for the work inside the ar71xx tree.

Signed-off-by: Mario Schroen <m.schroen@web.de>

dedeckeh commented:
The PR has formal issues as it has no SoB; please check https://openwrt.org/submitting-patches#submitting_patches

bluelineXY commented:
Hi, 
i add the sign off to the comment. 
If i rebase it i can resue the pull here and have to create a new one :/.

ynezz commented:
@bluelineXY Please don't, read https://openwrt.org/submitting-patches again and mainly focus on https://openwrt.org/submitting-patches#squashing_commits

bluelineXY commented:
Thx for the hint. Last time after rebase the PR was unusable

981213 commented:
Hi!

Last time I saw this PR you created a qca9533.dtsi while there is already a qca953x.dtsi available (qca953x.dtsi is suitable for qca9531 and qca9533).

And it seems that the dts is missing now :)

blocktrron commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -239,3 +239,21 @@ define Device/tplink_tl-wr2543-v1
+  KERNEL := kernel-bin | append-dtb | lzma | tplink-v1-header
+  IMAGES := sysupgrade.bin factory.bin
+  IMAGE/sysupgrade.bin := append-rootfs | tplink-safeloader sysupgrade | append-metadata | check-size $$$$(IMAGE_SIZE)
+  IMAGE/factory.bin := append-rootfs | tplink-safeloader factory
```
- `LOADER_TYPE` is only needed when we need to use an LZMA loader for the kernel. It looks like no such loader is used for this device, thus the line can be dropped

 - if you inherit `Device/tplink-safeloader` `KERNEL` and `IMAGE*` lines can be dropped

bluelineXY replied:
done

bluelineXY replied:
I had to change back the Config but i hope it fits now.

981213 replied:
As @blocktrron said you can reuse Device/tplink-safeloader template to reduce code duplication :)

blocktrron commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -197,6 +197,42 @@ static struct device_info boards[] = {
+			{"signature", 0x32000, 0x00400},
+			{"device-id", 0x33000, 0x00100},
+			{"os-image", 0x40000, 0x190000},
+			{"file-system", 0x1d0000, 0x5E0000},
```
Could you try if the device boots dynamic partitioning in place for kernel-image / filesystem? See this commit how to switch to dynamic partitioning: https://github.com/openwrt/openwrt/commit/7c78be1b747eb0c8d64da67deb3a8aec75bd7b9c

bluelineXY replied:
I will test the change - as i know i had this already running for testing. 

Do i need to add MTDPARTS into the Device definition or will the current setting in the DTS File be okay ? (there is a firmware partition)

bluelineXY replied:
Hi, 

it works with the current commit.

blocktrron commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -24,6 +24,7 @@ ath79_setup_interfaces()
 	tplink,tl-mr3040-v2|\
 	tplink,tl-wa901nd-v2|\
 	tplink,tl-wr703n|\
+	tplink,cpe210-v3|\
```
This should be ordered alphabetically.

bluelineXY replied:
done

blocktrron commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -54,6 +54,14 @@ glinet,gl-ar150)
 	ucidef_set_led_netdev "wan" "WAN" "$boardname:green:wan" "eth0"
 	ucidef_set_led_switch "lan" "LAN" "$boardname:green:lan" "switch0" "0x02"
 	;;
+tplink,cpe210-v3)
```
This should be ordered alphabetically.

bluelineXY replied:
done

blocktrron reviewed (commented):
As @981213 pointed out, the dts is currently missing. Quick look over the files currently present :smiley:

981213 commented on target/linux/ath79/dts/qca9533.dtsi:
```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
As I've said qca953x.dtsi is suitable for both qca9531 and qca9533. You don't need to create this file.
There are several mistakes in this file, but as it isn't needed in the first place I'm not going to point them all out :)

981213 commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,166 @@
+	compatible = "tplink,cpe210-v3", "qca,qca9533";
+	model = "TP-LINK CPE210 v3";
+
+	memory@0 {
```
Memory size can be auto-detected on ath79. This node isn't needed.

981213 commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,166 @@
+		lan {
+			label = "tp-link:green:lan";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
The default 'default-state' is 'off' so all the default-state lines can be dropped.

981213 commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,166 @@
+	};
+};
+
+&mdio0 {
```
phy4 is on builtin switch. It won't be accessible through mdio0.

981213 commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,166 @@
+	phy-handle = <&phy4>;
+	phy-mode = "mii";
+
+	fixed-link {
```
When you defined a 'fixed link' the phy-handle are ignored. eth0 is connected to phy4 on switch so the link status should be auto polled by using phy-handle instead of using a 'fixed-link'

981213 reviewed (changes requested):
Hi!
several things about the dts below :)

bluelineXY commented:
Hi, 
hopefully with this commit i got all the things managed.

981213 commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,139 @@
+	compatible = "tplink,cpe210-v3", "qca,qca9533";
+	model = "TP-LINK CPE210 v3";
+
+	extosc: ref {
```
This is already defined in qca953x.dtsi, so this node can be dropped.

bluelineXY replied:
done

981213 commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,139 @@
+	};
+};
+
+&pll {
```
also defined in qca953x.dtsi. This can be dropped too.

bluelineXY replied:
done

981213 commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -239,3 +239,13 @@ define Device/tplink_tl-wr2543-v1
+  DEVICE_TITLE := TP-Link CPE210 v3
+  DEVICE_PACKAGES := rssileds
+  TPLINK_BOARD_ID := CPE210V3
+  IMAGES := sysupgrade.bin factory.bin
```
This IMAGES line can be dropped. Device/tplink-safeloader includes Device/tplink, which has added factory.bin to IMAGES

bluelineXY replied:
done

981213 reviewed (changes requested):
Sorry, I missed 3 minor things last time :(

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,116 @@
+			};
+
+			partition@20000 {
+				label = "pation-table";
```
Is that a typo: "pation"?

bluelineXY replied:
jupp was a type and is fixed now

adrianschmutzler commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -197,6 +197,41 @@ static struct device_info boards[] = {
+			{NULL, 0, 0}
+		},
+
+		.first_sysupgrade_partition = "os-image",
```
I don't find "os-image" above. Am I getting something wrong?

bluelineXY replied:
thats a good one and i change this now to firmware 
BUT
if you have a look to other devices in this file, they have the same issue. 
Maybe this was overseen by the change to dynamic partitions in the patch mentioned in a comment  before

should this been changed as well - would make sense ?

adrianschmutzler replied:
Someone else has to judge on that, I'm not familiar with the underlying mechanics.

ynezz commented:
@bluelineXY please rebase on current master and retest
@adrianschmutzler @robimarko Can you please test this as well on your devices and provide me with your `Tested-by:` and boot log?

robimarko commented:
I cant test as I dont have CPE210 v3 at all.
My PR was based on testing by another user

adrianschmutzler commented:
I will receive a donated CPE210v3 soon and then can test both ath79 and ar71xx patches.
If other people want to test, I rebased both patches here (last two commits):
https://github.com/adrianschmutzler/openwrt/commits/cpev3all

adrianschmutzler commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -197,6 +197,41 @@ static struct device_info boards[] = {
+			{"device-info", 0x31400, 0x00400},
+			{"signature", 0x32000, 0x00400},
+			{"device-id", 0x33000, 0x00100},
+			{"firmware", 0x40000, 0x7B0000},
```
This should be:
{"firmware", 0x40000, 0x770000} ?

bluelineXY replied:
yup

bluelineXY replied:
i also will try to build the current version with all the latest commits i have done on my CPE210v3

bluelineXY replied:
Hi,
was trying to compile the version - but it ends with a core dump during build of the sysupgrade bin image. 
I had to change in tplink-safeloader.c the ".first_sysupgrade_partition" from firmware to os-partition. 

Did you have the same problem ? 
I saw this is set even on other devices, even they don't have a partition with this name. 
(e.g. ARCHER-A7-V5)

adrianschmutzler replied:
I did not try yet, but I think you are right. However, it is not "os-partition", but "os-image". Looks like this is already anticipating the splitting of "firmware" into "os-image" and "file-system".
I also do not really understand why sometimes the last partition is "support-list" and sometimes it is "file-system". However, I would stick to the other CPExxx which use "support-list" in this case.
If it works, you should again squash your commits into one ...

adrianschmutzler commented:
Was able to build with "os-image", device seemed to boot at least partially (not bootloop), but I was not able to ping or connect.
I do not have any means to investigate the boot process further.

bluelineXY commented:
That's strange did you start from the original factory software and use the tftp procedure to flash the openwrt factory.bin ?

I had problems using sysupgrade after changing something at the partition layout - ends in a boot loop. But usually even the sysupgrade package works for upgrades without changing the partition layout.

My CPE210v3 is running stable with the latest commit (which i also patched in the berlin freifunk sw).

adrianschmutzler commented:
Did something wrong during the build, will redo the images and test during the day.
However, from what the other devices did, I would additionally include the following into the target/linux/ath79/image/generic-tp-link.mk:

IMAGE_SIZE := 7680k
LOADER_TYPE := elf
KERNEL := kernel-bin | append-dtb | lzma | tplink-v1-header -O

If everything runs, I plan on opening a new PR where I want to:
- Provide ath79 support for CPE210v3
- Move ar71xx CPE210v2 to dynamic partitioning
- Provide ath79 support for CPE210v2
- Provide ar71xx support for CPE210v3

I would like to just include your patch into this series to make it more concise, if you do not object.

bluelineXY commented:
As i include the "Device/tplink-safeloader" the Kernel line is coming from there (without -O). 
Regarding Loader and Size i guess it's not relay relevant for the build process. 
For the IMAGE_SIZE i did a grep thru the whole source tree - no relevant match
For LOADER_TYPE i did the same - usage is quite minimal and more or less filenaming relevant. 

I would suggest to push this PR until this is merged as there is a lot of input and discussion happend here. 
what do you think ?

adrianschmutzler commented:
> As i include the "Device/tplink-safeloader" the Kernel line is coming from there (without -O).
I included it for the -O. This here sounds like we need it:
https://github.com/openwrt/openwrt/commit/0a2f21fa715bcf2e2975e32a1843a13fa088c640

> For the IMAGE_SIZE i did a grep thru the whole source tree - no relevant match
Well, thats true, maybe its just for the rare cases of "check-size $$$$(IMAGE_SIZE)"
However, the cpe210-v2 will then be the only device in generic-tp-link.mk without this defined. If it's not required, I can live with that.

> For LOADER_TYPE i did the same - usage is quite minimal and more or less filenaming relevant.
Yes. I still would include it for tidyness. however, I agree that it is not necessary.

>  would suggest to push this PR until this is merged as there is a lot of input and discussion happend here. what do you think ?
I can live with that. However, I would be glad if you squashed it so I can rebase my other stuff on it. 
Despite the stuff discussed above in this, IMHO the following still needs to be done:
- Change to os-image
- The entry in target/linux/ath79/image/generic-tp-link.mk needs to be moved (sorting is wrong since you changed from tl-cpe210-v2 to cpe210-v2)

Thanks for your work! Maybe we finally get this in.

adrianschmutzler commented:
I will build and test the following now on both devices:
https://github.com/adrianschmutzler/openwrt/commits/cpev23merge

................

adrianschmutzler commented:
Test successful for my image with the following changes:
os-image fix and IMAGE_SIZE/KERNEL/LOADER_TYPE added

However, I found that qca953x.dtsi introduces an eth1 which I suppose is not actually present on the device:
https://github.com/openwrt/openwrt/blob/cce55a17fc878bb1a6f84f7b2f105d34a7a5fe37/target/linux/ath79/dts/qca953x.dtsi#L273
How do we want to deal with that?

Bootlog:
[    0.000000] Linux version 4.14.107 (adsc@buildfff) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r9730-ae9d3a25c2)) #0 Mon Mar 25 21:50:28 2019
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019374 (MIPS 24Kc)
[    0.000000] MIPS: machine is TP-LINK CPE210 v3
[    0.000000] SoC: Qualcomm Atheros QCA9533 ver 2 rev 0
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000] free_area_init_node: node 0, pgdat 804ffab0, node_mem_map 81007060
[    0.000000]   Normal zone: 128 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x470 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,115200n8 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 58340K/65536K available (3778K kernel code, 153K rwdata, 808K rodata, 1204K init, 204K bss, 7196K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 650.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 5880801374 ns
[    0.000010] sched_clock: 32 bits at 325MHz, resolution 3ns, wraps every 6607641598ns
[    0.008274] Calibrating delay loop... 432.53 BogoMIPS (lpj=2162688)
[    0.074881] pid_max: default: 32768 minimum: 301
[    0.079973] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.086946] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.098946] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.109399] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.115940] pinctrl core: initialized pinctrl subsystem
[    0.122704] NET: Registered protocol family 16
[    0.158320] clocksource: Switched to clocksource MIPS
[    0.164959] NET: Registered protocol family 2
[    0.170527] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.177897] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.184691] TCP: Hash tables configured (established 1024 bind 1024)
[    0.191597] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.197782] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.204785] NET: Registered protocol family 1
[    0.209472] PCI: CLS 0 bytes, default 32
[    0.213526] Crashlog allocated RAM at address 0x3f00000
[    0.220825] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    0.233382] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.239600] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.261396] io scheduler noop registered
[    0.265536] io scheduler deadline registered (default)
[    0.273513] pinctrl-single 1804002c.pinmux: 576 pins at pa b804002c size 72
[    0.282103] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.289754] console [ttyS0] disabled
[    0.293584] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 8, base_baud = 1562500) is a 16550A
[    0.302755] console [ttyS0] enabled
[    0.310321] bootconsole [early0] disabled
[    0.335145] m25p80 spi0.0: s25fl064k (8192 Kbytes)
[    0.340232] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.346788] Creating 6 MTD partitions on "spi0.0":
[    0.351770] 0x000000000000-0x000000020000 : "u-boot"
[    0.357942] 0x000000020000-0x000000030000 : "partion-table"
[    0.364681] 0x000000030000-0x000000040000 : "product-info"
[    0.371365] 0x000000040000-0x0000007c0000 : "firmware"
[    0.378039] 2 tplink-fw partitions found on MTD device firmware
[    0.384246] Creating 2 MTD partitions on "firmware":
[    0.389396] 0x000000000000-0x000000184433 : "kernel"
[    0.395393] 0x000000190000-0x000000780000 : "rootfs"
[    0.401447] mtd: device 5 (rootfs) set to be root filesystem
[    0.407337] 1 squashfs-split partitions found on MTD device rootfs
[    0.413815] 0x000000420000-0x000000780000 : "rootfs_data"
[    0.420262] 0x0000007c0000-0x0000007f0000 : "config"
[    0.426240] 0x0000007f0000-0x000000800000 : "ART"
[    0.433485] libphy: Fixed MDIO Bus: probed
[    0.778772] libphy: ag71xx_mdio: probed
[    0.784212] libphy: ar8xxx-mdio: probed
[    0.797495] switch0: Atheros AR8229 rev. 1 switch registered on mdio-bus.0
[    1.190671] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd042, driver=Generic PHY]
[    1.201271] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    1.207845] ag71xx 1a000000.eth: invalid MAC address, using random address
[    1.550256] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.560340] eth1: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    1.568270] GACT probability on
[    1.571620] u32 classifier
[    1.574414]     input device check on
[    1.578185]     Actions configured
[    1.582837] NET: Registered protocol family 10
[    1.593590] Segment Routing with IPv6
[    1.597488] NET: Registered protocol family 17
[    1.602233] 8021q: 802.1Q VLAN Support v1.8
[    1.614875] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.632531] Freeing unused kernel memory: 1204K
[    1.637208] This architecture does not have kernel memory protection.
[    2.538333] random: fast init done
[    2.640519] init: Console is alive
[    2.644330] init: - watchdog -
...

adrianschmutzler commented:
Hmm, don't think the CPE210 v2/v3 has a switch either:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca953x.dtsi#L241

Looks like we need two dtsi here?

ynezz commented:
> Looks like we need two dtsi here?

Hm, I've looked at this quickly, but can't you just `&mdio1 { status = "disabled"; }` ?

adrianschmutzler commented:
> Hm, I've looked at this quickly, but can't you just &mdio1 { status = "disabled"; } ?

@ynezz Thanks. Will just try this. (I'm a DTS noob)

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,116 @@
+&eth0 {
+        status = "okay";
+        phy-handle = <&swphy4>;
+        mtd-mac-address = <&info 0x8>;
```
Those have leading spaces. All three lines eth0 ...

adrianschmutzler replied:
These leading spaces.

981213 commented:
> Hmm, don't think the CPE210 v2/v3 has a switch either:
> https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca953x.dtsi#L241
> 
> Looks like we need two dtsi here?

Add the following into your dts if you want to disable eth1:
```
&eth1 {
    compatible = "syscon", "simple-mfd";
}
```
The switch node in qca953x.dtsi is the builtin switch in qca953x soc. The ethernet PHY used by eth0 is part of that switch so the switch should be left enabled as-is for eth0 to work.

adrianschmutzler commented:
Thanks @981213 . Added the adjusted &eth1 as suggested and tested successfully.

I now tested my whole set of CPE210 v2/v3 patches located here successfully:

https://github.com/adrianschmutzler/openwrt/commits/cpev23merge6

Would be glad if we got this in before the branch-off.

@bluelineXY  I included a patch with all changes (also the KERNEL/LOADER etc. where we did not agree so far) in my series. How do you want to go on with this PR? Do you want to squash everything in your patch?

Bootlog for CPE210v3 @ ath79:
root@OpenWrt:~# dmesg
[    0.000000] Linux version 4.14.107 (adsc@buildfff) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r9730-ae9d3a25c2)) #0 Mon Mar 25 21:50:28 2019
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019374 (MIPS 24Kc)
[    0.000000] MIPS: machine is TP-LINK CPE210 v3
[    0.000000] SoC: Qualcomm Atheros QCA9533 ver 2 rev 0
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000] free_area_init_node: node 0, pgdat 804ffab0, node_mem_map 81007060
[    0.000000]   Normal zone: 128 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x470 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,115200n8 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 58340K/65536K available (3778K kernel code, 153K rwdata, 808K rodata, 1204K init, 204K bss, 7196K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 650.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 5880801374 ns
[    0.000010] sched_clock: 32 bits at 325MHz, resolution 3ns, wraps every 6607641598ns
[    0.008272] Calibrating delay loop... 432.53 BogoMIPS (lpj=2162688)
[    0.074881] pid_max: default: 32768 minimum: 301
[    0.079973] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.086946] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.098939] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.109390] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.115931] pinctrl core: initialized pinctrl subsystem
[    0.122699] NET: Registered protocol family 16
[    0.158293] clocksource: Switched to clocksource MIPS
[    0.164933] NET: Registered protocol family 2
[    0.170503] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.177873] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.184667] TCP: Hash tables configured (established 1024 bind 1024)
[    0.191570] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.197757] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.204759] NET: Registered protocol family 1
[    0.209447] PCI: CLS 0 bytes, default 32
[    0.213515] Crashlog allocated RAM at address 0x3f00000
[    0.220807] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    0.233351] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.239566] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.261388] io scheduler noop registered
[    0.265527] io scheduler deadline registered (default)
[    0.273502] pinctrl-single 1804002c.pinmux: 576 pins at pa b804002c size 72
[    0.282087] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.289737] console [ttyS0] disabled
[    0.293568] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 8, base_baud = 1562500) is a 16550A
[    0.302738] console [ttyS0] enabled
[    0.310305] bootconsole [early0] disabled
[    0.335128] m25p80 spi0.0: s25fl064k (8192 Kbytes)
[    0.340212] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.346771] Creating 6 MTD partitions on "spi0.0":
[    0.351754] 0x000000000000-0x000000020000 : "u-boot"
[    0.357930] 0x000000020000-0x000000030000 : "partion-table"
[    0.364662] 0x000000030000-0x000000040000 : "product-info"
[    0.371346] 0x000000040000-0x0000007c0000 : "firmware"
[    0.378013] 2 tplink-fw partitions found on MTD device firmware
[    0.384221] Creating 2 MTD partitions on "firmware":
[    0.389370] 0x000000000000-0x00000018442f : "kernel"
[    0.395366] 0x000000190000-0x000000780000 : "rootfs"
[    0.401425] mtd: device 5 (rootfs) set to be root filesystem
[    0.407321] 1 squashfs-split partitions found on MTD device rootfs
[    0.413796] 0x000000420000-0x000000780000 : "rootfs_data"
[    0.420244] 0x0000007c0000-0x0000007f0000 : "config"
[    0.426219] 0x0000007f0000-0x000000800000 : "ART"
[    0.433469] libphy: Fixed MDIO Bus: probed
[    0.778742] libphy: ag71xx_mdio: probed
[    0.784190] libphy: ar8xxx-mdio: probed
[    0.797457] switch0: Atheros AR8229 rev. 1 switch registered on mdio-bus.0
[    1.190640] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd042, driver=Generic PHY]
[    1.201248] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    1.209207] GACT probability on
[    1.212472] u32 classifier
[    1.215262]     input device check on
[    1.219111]     Actions configured
[    1.223719] NET: Registered protocol family 10
[    1.234094] Segment Routing with IPv6
[    1.237991] NET: Registered protocol family 17
[    1.242750] 8021q: 802.1Q VLAN Support v1.8
[    1.255359] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.273013] Freeing unused kernel memory: 1204K
[    1.277698] This architecture does not have kernel memory protection.
[    2.286713] init: Console is alive
[    2.290651] init: - watchdog -
[    2.498315] random: fast init done
...

bluelineXY commented:
nice one. 
I will do the test as well and integrate and squash then all into this pr

adrianschmutzler commented:
Interesting, "check-size $$$$(IMAGE_SIZE) " is present for ar71xx (https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/image/common-tp-link.mk#L76) and for individual devices in ath79 (https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/image/generic-tp-link.mk#L260), but not in the tplink-safeloader definition.

bluelineXY commented:
regarding the LOADER_TYPE i saw it's defined in Device/tplink which we use because of including the tplink-safeloader. 
So i guess it's set and fine and i squash the latest changes. 

Currnelty i don't know how to deal with the merge i  have here  
f81c8b3

adrianschmutzler commented:
> regarding the LOADER_TYPE i saw it's defined in Device/tplink which we use because of including the tplink-safeloader.

Yes, it is defined to "gz" there, where the CPE210 has used "elf" so far. It really seems to be only a labelling issue as you suggest, but I tend to be over-careful with those things.

> Currnelty i don't know how to deal with the merge i have here
Never merge when you do a PR. Just stack patches and then REBASE to the latest target. Since your changes after the rebase are small, I would recommend going back BEFORE the merge with e.g. "git reset --hard @~2", then rebase onto the current master, and then apply the remaining changes.

If you do not want to change anything individually, you can also remove your current branch with "git branch -D <name>", recreate it, use "git am" for my patches, and then "git push -f" to github again.
I do not care whose signed-off is below the patch, so if you want to take over/back the changes, feel free to do so. (It will still be easier than manually collecting everything ... ;-))

adrianschmutzler commented:
If the squash/rebase proves to be a bigger hold-up, you can also temporarily grant me push rights to your GitHub repo. Then we can still keep this PR ...

bluelineXY commented:
puh was bit tricky but i guess i made it ...

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts:
```diff
@@ -0,0 +1,119 @@
+			};
+
+			partition@20000 {
+				label = "partion-table";
```
partition-table. Also have that wrong in my version ...

adrianschmutzler commented:
Squash looks good. You still have the leading spaces in the DTS and are missing the KERNEL line with -O compared to my version. (I know that we disagreed on this, I just add this for completeness; if my tplink-safeloader Patch is accepted, it won't matter anyway).
And the partion-table made it through ...

adrianschmutzler commented:
Okay, better :-) Still leading spaces issue and technically, the SUPPORTED_DEVICES is not required since the cpe210-v3 for ar71xx is not merged YET.

adrianschmutzler commented:
... and BTW: Your whole commit message is indented. Some copy/paste thing?

adrianschmutzler commented:
Okay, do you want to push my other three patches here? Otherwise I would create a new PR for those (obviously including yours, as this one cannot be excluded).
My aim is to get them all in at once ...

adrianschmutzler commented:
This is tested and should be ready for reviewing/merging.

bluelineXY commented:
Hi, 
do you think that's a good idea ?
It maybe to large if we have all in one. should we maybe split it by arch ? 
So we put the v2 on ath79 also in this PR and the ar71xx in you new one ? 

i can merge them also here so we don't loose this  inputs here.

adrianschmutzler commented:
We cannot add the CPE210v2 to ath79 before ar71xx has moved to dynamic partitioning. (Unless you want to add it with fixed partition and then change both). tplink-safeloader.c is shared.

Despite that, your PR is separate and may also be merged without caring for the other stuff. However, I do hope that at some point, when someone has time for caring about device support, all might be merged in one shot.

adrianschmutzler commented:
Despite that, to be honest, I still hope for the ar71xx stuff being merged before the branch-off, so I will be included in the stable release.

apcameron commented:
@adrianschmutzler The last PR I submitted for ar71xx I was told to create it for ATH79 Instead.
See PR #1631

adrianschmutzler commented:
> The last PR I submitted for ar71xx I was told to create it for ATH79 Instead.

Well, it is understandable to force ath79 support for new devices. However, AFTER ath79 support is established, I don't see why support patches for ar71xx should generally be rejected. Especially, since in the CPE210v3 case the ar71xx patch is relatively simple (same hardware).
I admit that my chances are not high. But at least for openwrt-19.xx ath79 support won't be complete, so I will stick to ar71xx there, and thus put some effort in it.

rotanid commented:
that's also my understanding, new hardware to ar71xx only if ath79 support is also present.
so in this case, everything should be fine with #1951 .

bluelineXY commented:
ok. thne lets wait how the story continues and hopefully it will be merged soon.

adrianschmutzler commented:
@apcameron suggested to implement status LEDs, which we were missing. I did that and tested for both v2/v3 today. My ath79 v2 patch is updated, if you want to include that with the v3, you can use this patch:
https://github.com/adrianschmutzler/openwrt/commit/ba576b7345755f03b9785a670ffd38766787e0da
I also added the fix for the leading spaces still missing.

As I said, this is already tested also for the v3. :-)

adrianschmutzler commented:
Nice, thanks for pushing. So we are ready and have to wait again.

adrianschmutzler commented:
@bluelineXY We have been advised to include the "SUPPORTED_DEVICES += cpe210-v3" into this commit, as you intended earlier (https://github.com/openwrt/openwrt/pull/1951#issuecomment-478621883)

Please do so.

bluelineXY committed:
ath79: add TP-Link CPE210 v3

        Specifications:

            * SoC: Qualcomm Atheros QCA9533 (650MHz)
            * RAM: 64MB
            * Storage: 8 MB SPI NOR
            * Wireless: 2.4GHz N based built into SoC 2x2
            * Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

        Installation:
            Flash factory image through stock firmware WEB UI or TFTP
            To get to TFTP recovery just hold reset button while powering
            on for around 4-5 seconds and release.
            Rename factory image to recovery.bin
            Stock TFTP server IP:192.168.0.100
            Stock device TFTP adress:192.168.0.254

        Thanks to robimarko for the work inside the ar71xx tree.
        Thanks to adrianschmutzler for deep discussion and fixes.

Signed-off-by: Mario Schroen <m.schroen@web.de>

bluelineXY commented:
i have added the line

adrianschmutzler commented:
Despite that, we were supposed to create a dtsi for v2/v3, also as you already suggested ;-)

From my point of view, it would be more logical if you already created the dtsi here, instead of me moving everything from the v3.dts to a dtsi afterwards in my PR.
I would tend to still name it "qca9533_tplink_cpe210-v2.dtsi", since v2 is the first device "in line". But I will not insist in that one if you want to have it with "v3".

ynezz commented:
@adrianschmutzler Please just do all the requested changes in your PR, I'm going to close this one. Just make sure it works and that @bluelineXY agree with the changes and have tested them on his device(s) as well.

ynezz commented:
Otherwise I miss the point of the merged PR :)

adrianschmutzler commented:
@ynezz Thanks for the review and the advice.
@bluelineXY If you want to keep your sole Signed-off, you can push the changes here and will move them to my PR later. Sorry for the semi-hostile takeover ...

I will re-test ath79 for both devices later in the evening.

adrianschmutzler commented:
Pushed an updated version to the other PR:
https://github.com/adrianschmutzler/openwrt/commits/cpev23merge6

bluelineXY commented:
i would suggest to name the file "qca9533_tplink_cpe210.dts"

because even the -v1 should be named like this with a -v1.
the question would be how the v1 realy differ beside the 2nd LAN Port. 

do you have the chance to get a -v1 device ?

ynezz commented:
> I would suggest to name the file "qca9533_tplink_cpe210.dts"

I've done this already in my tree. Thanks for the hint.

bluelineXY commented:
thats a good one - i have done this now.

adrianschmutzler commented:
> do you have the chance to get a -v1 device ?

I can ask for one after I have sent back my borrowed v3 :-)

adrianschmutzler commented:
> do you have the chance to get a -v1 device ?

I've had a look onto the WIP support patch for the CPE510v1 (https://github.com/openwrt/openwrt/pull/1704). The changes are beyond what I can do at the moment, both because of my very limited DTS understanding and since I have no serial console or similar. Despite that, for CPE210v1 one has to deal with the additional OKLI loader.

I will have another look when the CPE510 patch is finished, but most probably the CPE210v1 will have to be someone else's job ....
